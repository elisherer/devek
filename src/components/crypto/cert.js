const printHex = (array, width = -1, indent = 0) => {
	if (!Array.isArray(array)) return `${array} (0x${array.toString(16)})`;
	let output = array
		.map(x => {
			const byte = x.toString(16);
			return byte.length === 2 ? byte : "0" + byte;
		})
		.join(":");
	if (width === -1 || output.length < width * 3) return output;
	return output
		.match(new RegExp(`.{1,${width * 3}}`, "g"))
		.join("\n" + " ".repeat(indent));
};

const openSslAliases = {
	countryName: "C",
	stateOrProvinceName: "ST",
	localityName: "L",
	organizationName: "O",
	organizationalUnitName: "OU",
	commonName: "CN",
	pkcs9_emailAddress: "emailAddress"
};

function replacer(key, value) {
	if (Array.isArray(value) && typeof value[0] === "number") {
		return printHex(value);
	}
	return value;
}

export const prettyCert = cert => {
	const publicKey =
		cert.subjectPublicKeyInfo.algorithm === "rsaEncryption"
			? `
                Public-Key: ${cert.subjectPublicKeyInfo.publicKey.publicKey}
                Modulus:
                    ${printHex(
											cert.subjectPublicKeyInfo.publicKey.modulus,
											15,
											20
										)}
                Exponent: ${printHex(
									cert.subjectPublicKeyInfo.publicKey.exponent
								)}`
			: `
                Public-Key: ${cert.subjectPublicKeyInfo.publicKey.publicKey}
                pub:
                    ${printHex(cert.subjectPublicKeyInfo.publicKey.pub, 15, 20)}
                ASN1 OID: ${cert.subjectPublicKeyInfo.publicKey.asn1Oid}
                NIST CURVE: ${cert.subjectPublicKeyInfo.publicKey.nistCurve}`;

	let extensions = "";
	if (cert.extensions) {
		extensions = `
        X509v3 extensions:\n`;
		Object.keys(cert.extensions).forEach(ext => {
			extensions += `            ${ext}: ${
				cert.extensions[ext].critical ? "critical" : ""
			}
                ${JSON.stringify(
									cert.extensions[ext].value,
									replacer,
									2
								).replace(/\n/g, "\n                ")}\n`;
		});
	}

	return `Certificate:
    Data:
        Version: ${cert.version + 1} (0x${cert.version})
        Serial Number:
            ${printHex(cert.serialNumber)}
    Signature Algorithm: ${cert.signature.algorithm}
        Issuer: ${Object.keys(cert.issuer)
					.map(d => (openSslAliases[d] || d) + "=" + cert.issuer[d])
					.join(", ")}
        Validity
            Not Before: ${cert.validity.notBefore.toGMTString()}
            Not After : ${cert.validity.notAfter.toGMTString()}
        Subject: ${Object.keys(cert.subject)
					.map(d => (openSslAliases[d] || d) + "=" + cert.subject[d])
					.join(", ")}
        Subject Public Key Info:
            Public Key Algorithm: ${
							cert.subjectPublicKeyInfo.algorithm
						}${publicKey}${extensions}
    Signature Algorithm: ${cert.signatureAlgorithm.algorithm}
         ${printHex(cert.signatureValue, 18, 9)}`;
};
