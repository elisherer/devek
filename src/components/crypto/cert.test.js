import { parseCertificate } from "./cert";

const c0_ec = `-----BEGIN CERTIFICATE-----
MIIBCTCBsAIJAI1GxVa/1o2pMAoGCCqGSM49BAMCMA0xCzAJBgNVBAMMAkNBMB4X
DTE5MDYxMTEzMjY1MloXDTE5MDcxMTEzMjY1MlowDTELMAkGA1UEAwwCRUUwWTAT
BgcqhkjOPQIBBggqhkjOPQMBBwNCAATCMoLhT+10QEHpcbu6k+koYhLBeEy8bUIH
RBXo0W29fDUDCRHus3NixkbxcarGxaLmhvwYFgQdbgaGjKx8IJLzMAoGCCqGSM49
BAMCA0gAMEUCIGISk7evLWaZKX7nANauKrJLTRCIbdB9O/CNNvGqYwAVAiEA9iNY
Vi+TkGrcl2TUe4a0xM0vTunzBgV+/jSoJsSWGkM=
-----END CERTIFICATE----- `;

const c1 = `MIIGazCCBFOgAwIBAgIISUvdceh9stEwDQYJKoZIhvcNAQELBQAwgZQxCzAJBgNV
BAYTAklUMQ4wDAYDVQQIEwVJdGFseTEOMAwGA1UEBxMFTWlsYW4xETAPBgNVBAoT
CENoZUJhbmNhMRIwEAYDVQQLEwlTaWN1cmV6emExGDAWBgNVBAMTD0RFViBDaGVC
YW5jYSBDQTEkMCIGCSqGSIb3DQEJARYVc2ljdXJlenphQGNoZWJhbmNhLml0MB4X
DTE5MDMwNTE0MDcwMFoXDTI4MDMwNTE0MDcwMFowgYcxGDAWBgNVBGEMD1BTRElU
LUNCLTAwMDAwMDELMAkGA1UEBgwCSVQxFzAVBgNVBAMMDkNCIFRQUCBTQU5EQk9Y
MQwwCgYDVQQLDANUUFAxFzAVBgNVBAoMDkNCIFRQUCBTQU5EQk9YMQ4wDAYDVQQI
DAVJdGFseTEOMAwGA1UEBwwFTWlsYW4wggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAw
ggIKAoICAQC3doBPCwtOlHDBhFDBdEM5b8bQgCJnAsiz0qdFWxkTuWTG6lXA6zTh
RUUqVCnb0pNVQt9o96lJ+DT/htKfxQTCIIMqVi5bgXTD2ZBJRBOOWV4avEG9J0LR
KVzxmN+s4DQvcCKSGZ6dNelRFhtHi6xIuVqvp6ldHdE1MlH+osP1A1PeFJ4wuaLz
zatO3rI/f+UaAuIb0y4sGdxHODcP/tuHCYxi55kAV+5BxntjJNO1FNFINvjNH1hC
+Kxyjn7bGgTxaGqo9AplXe653BEQsPhOdh12E3hR5FRuqqGzcVpDQo/Q/xtaaV/Z
Qeqck4gdV0Nn8HpOuoq+Ssizolyi0cMyyQzhERHFtDy4Zs5QM51vica+hrMpJQhE
Pda7ZVXLewTBSC2XowdAtbFmILnvk6/AkbxFOWx7ESA3SiNixv9Slh5F0g9Ste8C
4WgasOZwAt2a3gKYxM//uVvyZEBzi2dg5Wjv7yfo/fnSFrRS3asi153KaHMP4CKy
NyQ7ZCDXHsAdBrbBXUvpopAHfgxNXNh6mSW8X10NFgF+yrIlnvOL67juvv4y5GLs
9Ee/T4GriCi0KBwqBp1EcGIVO8JuFlMEQmyBIUmiX44Smv+zl0+ptqQWvdJDy+hw
ieBd6tbBUr+hhzlZCrXU9EKIJDrjZzSXN1JLRK1E3QmLBCg2M/4pCQIDAQABo4HL
MIHIMAsGA1UdDwQEAwIBxjAdBgNVHSUEFjAUBggrBgEFBQcDAgYIKwYBBQUHAwEw
EQYDVR0RBAowCIIGdHBwLml0MIGGBggrBgEFBQcBAwR6MHgwCAYGBACORgEBMAsG
BgQAjkYBAwIBFDAIBgYEAI5GAQQwEwYGBACORgEGMAkGBwQAjkYBBgMwQAYGBACB
mCcCMDYwJjARBgcEAIGYJwEDDAZQU1BfQUkwEQYHBACBmCcBAgwGUFNQX1BJDAhD
aGVCYW5jYQwCQ0IwDQYJKoZIhvcNAQELBQADggIBAGA6JHa3E5+xgZAWMUO5Mg0u
0tKTSLFCjGhnhORvqlSYDtFVYRYixFARxvJrn2rtBCMl+1VIq2x0ZrkPQ03s/10w
jhrKuJ0yDX9mgxWDNGdDq/QXLOnCm7Ly0egOqb4a0TkvJJToCL74y55Xj7CPgGxp
lQjnsOesCc/Z/SJPB5gHRQ8PClYoHZy5qWtHWXeRNzRZGruU30sHYgmWrajwtWKy
FuPbhpN7Sx9L/LsG5SQj8s5R6Xqv8pe6RrJCiOmjXsnlIDpmu7L5WlldovNGubAt
b+EqtNMiLzsuguUNWsFjzBn2Hu5PhQ1DH36kRdPRbHK3Tcm/kMXLKwxCFIN9CA4v
O9PE7YZD6Hlvx3xqsbVgZkFrFHjQSgCNyHraH/gbZIwQ/UNkFxs2e0Pfcraegbhu
JarXqDVJnNVM6Kcoqm24ppG05bjzNwd8uPmL50ceD1cxmn54Dro34kJPbuQ/ZrEt
m+XHiiv4rQyIJKvEmeZeK+44q5IcFmBLKhDn1Tc9PR+iB6encnQQtFtnr5NwyLjV
mHiiX8uciIDSXOfqW9toLLKlJ2KII+o4Qhe0H1mKSTZ2ecgS+IT4HHQZEW+DMQ/I
MKh4pIAOasqKbd53deOvUSpSSnhORzVLJwQLhg8VmRzSR5T9V13ad/ptbv37dBFp
P0VbDxUMTGrdMmK2BJuq`;

//eslint-disable-next-line
const c2 = `MIIC2jCCAkMCAg38MA0GCSqGSIb3DQEBBQUAMIGbMQswCQYDVQQGEwJKUDEOMAwG
A1UECBMFVG9reW8xEDAOBgNVBAcTB0NodW8ta3UxETAPBgNVBAoTCEZyYW5rNERE
MRgwFgYDVQQLEw9XZWJDZXJ0IFN1cHBvcnQxGDAWBgNVBAMTD0ZyYW5rNEREIFdl
YiBDQTEjMCEGCSqGSIb3DQEJARYUc3VwcG9ydEBmcmFuazRkZC5jb20wHhcNMTIw
ODIyMDUyNzQxWhcNMTcwODIxMDUyNzQxWjBKMQswCQYDVQQGEwJKUDEOMAwGA1UE
CAwFVG9reW8xETAPBgNVBAoMCEZyYW5rNEREMRgwFgYDVQQDDA93d3cuZXhhbXBs
ZS5jb20wggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQC0z9FeMynsC8+u
dvX+LciZxnh5uRj4C9S6tNeeAlIGCfQYk0zUcNFCoCkTknNQd/YEiawDLNbxBqut
bMDZ1aarys1a0lYmUeVLCIqvzBkPJTSQsCopQQ9V8WuT252zzNzs68dVGNdCJd5J
NRQykpwexmnjPPv0mvj7i8XgG379TyW6P+WWV5okeUkXJ9eJS2ouDYdR2SM9BoVW
+FgxDu6BmXhozW5EfsnajFp7HL8kQClI0QOc79yuKl3492rH6bzFsFn2lfwWy9ic
7cP8EpCTeFp1tFaD+vxBhPZkeTQ1HKx6hQ5zeHIB5ySJJZ7af2W8r4eTGYzbdRW2
4DDHCPhZAgMBAAEwDQYJKoZIhvcNAQEFBQADgYEAQMv+BFvGdMVzkQaQ3/+2noVz
/uAKbzpEL8xTcxYyP3lkOeh4FoxiSWqy5pGFALdPONoDuYFpLhjJSZaEwuvjI/Tr
rGhLV1pRG9frwDFshqD2Vaj4ENBCBh6UpeBop5+285zQ4SI7q4U9oSebUDJiuOx6
+tZ9KynmrbJpTSi0+BM=`;

describe("crypto/cert", () => {
	test("cert 1", () => {
		const { buffer, ...asn } = parseCertificate(c0_ec); // eslint-disable-line no-unused-vars
		expect(asn).toMatchSnapshot();
	});

	test("cert 2", () => {
		const { buffer, ...asn } = parseCertificate(c1); // eslint-disable-line no-unused-vars
		expect(asn).toMatchSnapshot();
	});

	test("cert 3", () => {
		const { buffer, ...asn } = parseCertificate(c2); // eslint-disable-line no-unused-vars
		expect(asn).toMatchSnapshot();
	});
});
