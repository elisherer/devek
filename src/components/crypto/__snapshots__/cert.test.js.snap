// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crypto/cert cert 1 1`] = `
Object {
  "extensions": undefined,
  "issuer": Object {
    "commonName": "CA",
  },
  "issuerUniqueID": undefined,
  "serialNumber": Array [
    141,
    70,
    197,
    86,
    191,
    214,
    141,
    169,
  ],
  "signature": Object {
    "algorithm": "ecdsa_with_SHA256",
    "parameters": undefined,
  },
  "signatureAlgorithm": Object {
    "algorithm": "ecdsa_with_SHA256",
    "parameters": undefined,
  },
  "signatureValue": Array [
    48,
    69,
    2,
    32,
    98,
    18,
    147,
    183,
    175,
    45,
    102,
    153,
    41,
    126,
    231,
    0,
    214,
    174,
    42,
    178,
    75,
    77,
    16,
    136,
    109,
    208,
    125,
    59,
    240,
    141,
    54,
    241,
    170,
    99,
    0,
    21,
    2,
    33,
    0,
    246,
    35,
    88,
    86,
    47,
    147,
    144,
    106,
    220,
    151,
    100,
    212,
    123,
    134,
    180,
    196,
    205,
    47,
    78,
    233,
    243,
    6,
    5,
    126,
    254,
    52,
    168,
    38,
    196,
    150,
    26,
    67,
  ],
  "subject": Object {
    "commonName": "EE",
  },
  "subjectPublicKeyInfo": Object {
    "algorithm": "ecPublicKey",
    "publicKey": Object {
      "asn1Oid": "X9_62_prime256v1",
      "nistCurve": "P-256",
      "pub": Array [
        4,
        194,
        50,
        130,
        225,
        79,
        237,
        116,
        64,
        65,
        233,
        113,
        187,
        186,
        147,
        233,
        40,
        98,
        18,
        193,
        120,
        76,
        188,
        109,
        66,
        7,
        68,
        21,
        232,
        209,
        109,
        189,
        124,
        53,
        3,
        9,
        17,
        238,
        179,
        115,
        98,
        198,
        70,
        241,
        113,
        170,
        198,
        197,
        162,
        230,
        134,
        252,
        24,
        22,
        4,
        29,
        110,
        6,
        134,
        140,
        172,
        124,
        32,
        146,
        243,
      ],
      "publicKey": "(256 bit)",
    },
  },
  "subjectUniqueID": undefined,
  "validity": Object {
    "notAfter": 2019-07-11T13:26:52.000Z,
    "notBefore": 2019-06-11T13:26:52.000Z,
  },
  "version": 0,
}
`;

exports[`crypto/cert cert 2 1`] = `
Object {
  "extensions": Object {
    "extKeyUsage": Object {
      "critical": false,
      "value": Array [
        "clientAuth",
        "serverAuth",
      ],
    },
    "keyUsage": Object {
      "critical": false,
      "value": Array [
        "digitalSignature",
        "nonRepudiation",
        "keyCertSign",
        "cRLSign",
      ],
    },
    "qcStatements": Object {
      "critical": false,
      "value": Object {
        "PSD2 qcStatement": Object {
          "ncaId": "CB",
          "ncaName": "CheBanca",
          "rolesOfPSP": Object {
            "PSP_AI": "PSP_AI",
            "PSP_PI": "PSP_PI",
          },
        },
        "QcCompliance": true,
        "QcRetentionPeriod": 20,
        "QcSSCD": true,
        "QcType": Array [
          "web",
        ],
      },
    },
    "subjectAltName": Object {
      "critical": false,
      "value": Object {
        "dNSName": "tpp.it",
      },
    },
  },
  "issuer": Object {
    "commonName": "DEV CheBanca CA",
    "countryName": "IT",
    "emailAddress": "sicurezza@chebanca.it",
    "localityName": "Milan",
    "organizationName": "CheBanca",
    "organizationalUnitName": "Sicurezza",
    "stateOrProvinceName": "Italy",
  },
  "issuerUniqueID": undefined,
  "serialNumber": Array [
    73,
    75,
    221,
    113,
    232,
    125,
    178,
    209,
  ],
  "signature": Object {
    "algorithm": "sha256WithRSAEncryption",
    "parameters": null,
  },
  "signatureAlgorithm": Object {
    "algorithm": "sha256WithRSAEncryption",
    "parameters": null,
  },
  "signatureValue": Array [
    96,
    58,
    36,
    118,
    183,
    19,
    159,
    177,
    129,
    144,
    22,
    49,
    67,
    185,
    50,
    13,
    46,
    210,
    210,
    147,
    72,
    177,
    66,
    140,
    104,
    103,
    132,
    228,
    111,
    170,
    84,
    152,
    14,
    209,
    85,
    97,
    22,
    34,
    196,
    80,
    17,
    198,
    242,
    107,
    159,
    106,
    237,
    4,
    35,
    37,
    251,
    85,
    72,
    171,
    108,
    116,
    102,
    185,
    15,
    67,
    77,
    236,
    255,
    93,
    48,
    142,
    26,
    202,
    184,
    157,
    50,
    13,
    127,
    102,
    131,
    21,
    131,
    52,
    103,
    67,
    171,
    244,
    23,
    44,
    233,
    194,
    155,
    178,
    242,
    209,
    232,
    14,
    169,
    190,
    26,
    209,
    57,
    47,
    36,
    148,
    232,
    8,
    190,
    248,
    203,
    158,
    87,
    143,
    176,
    143,
    128,
    108,
    105,
    149,
    8,
    231,
    176,
    231,
    172,
    9,
    207,
    217,
    253,
    34,
    79,
    7,
    152,
    7,
    69,
    15,
    15,
    10,
    86,
    40,
    29,
    156,
    185,
    169,
    107,
    71,
    89,
    119,
    145,
    55,
    52,
    89,
    26,
    187,
    148,
    223,
    75,
    7,
    98,
    9,
    150,
    173,
    168,
    240,
    181,
    98,
    178,
    22,
    227,
    219,
    134,
    147,
    123,
    75,
    31,
    75,
    252,
    187,
    6,
    229,
    36,
    35,
    242,
    206,
    81,
    233,
    122,
    175,
    242,
    151,
    186,
    70,
    178,
    66,
    136,
    233,
    163,
    94,
    201,
    229,
    32,
    58,
    102,
    187,
    178,
    249,
    90,
    89,
    93,
    162,
    243,
    70,
    185,
    176,
    45,
    111,
    225,
    42,
    180,
    211,
    34,
    47,
    59,
    46,
    130,
    229,
    13,
    90,
    193,
    99,
    204,
    25,
    246,
    30,
    238,
    79,
    133,
    13,
    67,
    31,
    126,
    164,
    69,
    211,
    209,
    108,
    114,
    183,
    77,
    201,
    191,
    144,
    197,
    203,
    43,
    12,
    66,
    20,
    131,
    125,
    8,
    14,
    47,
    59,
    211,
    196,
    237,
    134,
    67,
    232,
    121,
    111,
    199,
    124,
    106,
    177,
    181,
    96,
    102,
    65,
    107,
    20,
    120,
    208,
    74,
    0,
    141,
    200,
    122,
    218,
    31,
    248,
    27,
    100,
    140,
    16,
    253,
    67,
    100,
    23,
    27,
    54,
    123,
    67,
    223,
    114,
    182,
    158,
    129,
    184,
    110,
    37,
    170,
    215,
    168,
    53,
    73,
    156,
    213,
    76,
    232,
    167,
    40,
    170,
    109,
    184,
    166,
    145,
    180,
    229,
    184,
    243,
    55,
    7,
    124,
    184,
    249,
    139,
    231,
    71,
    30,
    15,
    87,
    49,
    154,
    126,
    120,
    14,
    186,
    55,
    226,
    66,
    79,
    110,
    228,
    63,
    102,
    177,
    45,
    155,
    229,
    199,
    138,
    43,
    248,
    173,
    12,
    136,
    36,
    171,
    196,
    153,
    230,
    94,
    43,
    238,
    56,
    171,
    146,
    28,
    22,
    96,
    75,
    42,
    16,
    231,
    213,
    55,
    61,
    61,
    31,
    162,
    7,
    167,
    167,
    114,
    116,
    16,
    180,
    91,
    103,
    175,
    147,
    112,
    200,
    184,
    213,
    152,
    120,
    162,
    95,
    203,
    156,
    136,
    128,
    210,
    92,
    231,
    234,
    91,
    219,
    104,
    44,
    178,
    165,
    39,
    98,
    136,
    35,
    234,
    56,
    66,
    23,
    180,
    31,
    89,
    138,
    73,
    54,
    118,
    121,
    200,
    18,
    248,
    132,
    248,
    28,
    116,
    25,
    17,
    111,
    131,
    49,
    15,
    200,
    48,
    168,
    120,
    164,
    128,
    14,
    106,
    202,
    138,
    109,
    222,
    119,
    117,
    227,
    175,
    81,
    42,
    82,
    74,
    120,
    78,
    71,
    53,
    75,
    39,
    4,
    11,
    134,
    15,
    21,
    153,
    28,
    210,
    71,
    148,
    253,
    87,
    93,
    218,
    119,
    250,
    109,
    110,
    253,
    251,
    116,
    17,
    105,
    63,
    69,
    91,
    15,
    21,
    12,
    76,
    106,
    221,
    50,
    98,
    182,
    4,
    155,
    170,
  ],
  "subject": Object {
    "commonName": "CB TPP SANDBOX",
    "countryName": "IT",
    "localityName": "Milan",
    "organizationIdentifier": "PSDIT-CB-000000",
    "organizationName": "CB TPP SANDBOX",
    "organizationalUnitName": "TPP",
    "stateOrProvinceName": "Italy",
  },
  "subjectPublicKeyInfo": Object {
    "algorithm": "rsaEncryption",
    "publicKey": Object {
      "exponent": 65537,
      "modulus": Array [
        0,
        183,
        118,
        128,
        79,
        11,
        11,
        78,
        148,
        112,
        193,
        132,
        80,
        193,
        116,
        67,
        57,
        111,
        198,
        208,
        128,
        34,
        103,
        2,
        200,
        179,
        210,
        167,
        69,
        91,
        25,
        19,
        185,
        100,
        198,
        234,
        85,
        192,
        235,
        52,
        225,
        69,
        69,
        42,
        84,
        41,
        219,
        210,
        147,
        85,
        66,
        223,
        104,
        247,
        169,
        73,
        248,
        52,
        255,
        134,
        210,
        159,
        197,
        4,
        194,
        32,
        131,
        42,
        86,
        46,
        91,
        129,
        116,
        195,
        217,
        144,
        73,
        68,
        19,
        142,
        89,
        94,
        26,
        188,
        65,
        189,
        39,
        66,
        209,
        41,
        92,
        241,
        152,
        223,
        172,
        224,
        52,
        47,
        112,
        34,
        146,
        25,
        158,
        157,
        53,
        233,
        81,
        22,
        27,
        71,
        139,
        172,
        72,
        185,
        90,
        175,
        167,
        169,
        93,
        29,
        209,
        53,
        50,
        81,
        254,
        162,
        195,
        245,
        3,
        83,
        222,
        20,
        158,
        48,
        185,
        162,
        243,
        205,
        171,
        78,
        222,
        178,
        63,
        127,
        229,
        26,
        2,
        226,
        27,
        211,
        46,
        44,
        25,
        220,
        71,
        56,
        55,
        15,
        254,
        219,
        135,
        9,
        140,
        98,
        231,
        153,
        0,
        87,
        238,
        65,
        198,
        123,
        99,
        36,
        211,
        181,
        20,
        209,
        72,
        54,
        248,
        205,
        31,
        88,
        66,
        248,
        172,
        114,
        142,
        126,
        219,
        26,
        4,
        241,
        104,
        106,
        168,
        244,
        10,
        101,
        93,
        238,
        185,
        220,
        17,
        16,
        176,
        248,
        78,
        118,
        29,
        118,
        19,
        120,
        81,
        228,
        84,
        110,
        170,
        161,
        179,
        113,
        90,
        67,
        66,
        143,
        208,
        255,
        27,
        90,
        105,
        95,
        217,
        65,
        234,
        156,
        147,
        136,
        29,
        87,
        67,
        103,
        240,
        122,
        78,
        186,
        138,
        190,
        74,
        200,
        179,
        162,
        92,
        162,
        209,
        195,
        50,
        201,
        12,
        225,
        17,
        17,
        197,
        180,
        60,
        184,
        102,
        206,
        80,
        51,
        157,
        111,
        137,
        198,
        190,
        134,
        179,
        41,
        37,
        8,
        68,
        61,
        214,
        187,
        101,
        85,
        203,
        123,
        4,
        193,
        72,
        45,
        151,
        163,
        7,
        64,
        181,
        177,
        102,
        32,
        185,
        239,
        147,
        175,
        192,
        145,
        188,
        69,
        57,
        108,
        123,
        17,
        32,
        55,
        74,
        35,
        98,
        198,
        255,
        82,
        150,
        30,
        69,
        210,
        15,
        82,
        181,
        239,
        2,
        225,
        104,
        26,
        176,
        230,
        112,
        2,
        221,
        154,
        222,
        2,
        152,
        196,
        207,
        255,
        185,
        91,
        242,
        100,
        64,
        115,
        139,
        103,
        96,
        229,
        104,
        239,
        239,
        39,
        232,
        253,
        249,
        210,
        22,
        180,
        82,
        221,
        171,
        34,
        215,
        157,
        202,
        104,
        115,
        15,
        224,
        34,
        178,
        55,
        36,
        59,
        100,
        32,
        215,
        30,
        192,
        29,
        6,
        182,
        193,
        93,
        75,
        233,
        162,
        144,
        7,
        126,
        12,
        77,
        92,
        216,
        122,
        153,
        37,
        188,
        95,
        93,
        13,
        22,
        1,
        126,
        202,
        178,
        37,
        158,
        243,
        139,
        235,
        184,
        238,
        190,
        254,
        50,
        228,
        98,
        236,
        244,
        71,
        191,
        79,
        129,
        171,
        136,
        40,
        180,
        40,
        28,
        42,
        6,
        157,
        68,
        112,
        98,
        21,
        59,
        194,
        110,
        22,
        83,
        4,
        66,
        108,
        129,
        33,
        73,
        162,
        95,
        142,
        18,
        154,
        255,
        179,
        151,
        79,
        169,
        182,
        164,
        22,
        189,
        210,
        67,
        203,
        232,
        112,
        137,
        224,
        93,
        234,
        214,
        193,
        82,
        191,
        161,
        135,
        57,
        89,
        10,
        181,
        212,
        244,
        66,
        136,
        36,
        58,
        227,
        103,
        52,
        151,
        55,
        82,
        75,
        68,
        173,
        68,
        221,
        9,
        139,
        4,
        40,
        54,
        51,
        254,
        41,
        9,
      ],
      "publicKey": "(4096 bit)",
    },
  },
  "subjectUniqueID": undefined,
  "validity": Object {
    "notAfter": 2028-03-05T14:07:00.000Z,
    "notBefore": 2019-03-05T14:07:00.000Z,
  },
  "version": 2,
}
`;

exports[`crypto/cert cert 3 1`] = `
Object {
  "extensions": undefined,
  "issuer": Object {
    "commonName": "Frank4DD Web CA",
    "countryName": "JP",
    "emailAddress": "support@frank4dd.com",
    "localityName": "Chuo-ku",
    "organizationName": "Frank4DD",
    "organizationalUnitName": "WebCert Support",
    "stateOrProvinceName": "Tokyo",
  },
  "issuerUniqueID": undefined,
  "serialNumber": 3580,
  "signature": Object {
    "algorithm": "sha1withRSAEncryption",
    "parameters": null,
  },
  "signatureAlgorithm": Object {
    "algorithm": "sha1withRSAEncryption",
    "parameters": null,
  },
  "signatureValue": Array [
    64,
    203,
    254,
    4,
    91,
    198,
    116,
    197,
    115,
    145,
    6,
    144,
    223,
    255,
    182,
    158,
    133,
    115,
    254,
    224,
    10,
    111,
    58,
    68,
    47,
    204,
    83,
    115,
    22,
    50,
    63,
    121,
    100,
    57,
    232,
    120,
    22,
    140,
    98,
    73,
    106,
    178,
    230,
    145,
    133,
    0,
    183,
    79,
    56,
    218,
    3,
    185,
    129,
    105,
    46,
    24,
    201,
    73,
    150,
    132,
    194,
    235,
    227,
    35,
    244,
    235,
    172,
    104,
    75,
    87,
    90,
    81,
    27,
    215,
    235,
    192,
    49,
    108,
    134,
    160,
    246,
    85,
    168,
    248,
    16,
    208,
    66,
    6,
    30,
    148,
    165,
    224,
    104,
    167,
    159,
    182,
    243,
    156,
    208,
    225,
    34,
    59,
    171,
    133,
    61,
    161,
    39,
    155,
    80,
    50,
    98,
    184,
    236,
    122,
    250,
    214,
    125,
    43,
    41,
    230,
    173,
    178,
    105,
    77,
    40,
    180,
    248,
    19,
  ],
  "subject": Object {
    "commonName": "www.example.com",
    "countryName": "JP",
    "organizationName": "Frank4DD",
    "stateOrProvinceName": "Tokyo",
  },
  "subjectPublicKeyInfo": Object {
    "algorithm": "rsaEncryption",
    "publicKey": Object {
      "exponent": 65537,
      "modulus": Array [
        0,
        180,
        207,
        209,
        94,
        51,
        41,
        236,
        11,
        207,
        174,
        118,
        245,
        254,
        45,
        200,
        153,
        198,
        120,
        121,
        185,
        24,
        248,
        11,
        212,
        186,
        180,
        215,
        158,
        2,
        82,
        6,
        9,
        244,
        24,
        147,
        76,
        212,
        112,
        209,
        66,
        160,
        41,
        19,
        146,
        115,
        80,
        119,
        246,
        4,
        137,
        172,
        3,
        44,
        214,
        241,
        6,
        171,
        173,
        108,
        192,
        217,
        213,
        166,
        171,
        202,
        205,
        90,
        210,
        86,
        38,
        81,
        229,
        75,
        8,
        138,
        175,
        204,
        25,
        15,
        37,
        52,
        144,
        176,
        42,
        41,
        65,
        15,
        85,
        241,
        107,
        147,
        219,
        157,
        179,
        204,
        220,
        236,
        235,
        199,
        85,
        24,
        215,
        66,
        37,
        222,
        73,
        53,
        20,
        50,
        146,
        156,
        30,
        198,
        105,
        227,
        60,
        251,
        244,
        154,
        248,
        251,
        139,
        197,
        224,
        27,
        126,
        253,
        79,
        37,
        186,
        63,
        229,
        150,
        87,
        154,
        36,
        121,
        73,
        23,
        39,
        215,
        137,
        75,
        106,
        46,
        13,
        135,
        81,
        217,
        35,
        61,
        6,
        133,
        86,
        248,
        88,
        49,
        14,
        238,
        129,
        153,
        120,
        104,
        205,
        110,
        68,
        126,
        201,
        218,
        140,
        90,
        123,
        28,
        191,
        36,
        64,
        41,
        72,
        209,
        3,
        156,
        239,
        220,
        174,
        42,
        93,
        248,
        247,
        106,
        199,
        233,
        188,
        197,
        176,
        89,
        246,
        149,
        252,
        22,
        203,
        216,
        156,
        237,
        195,
        252,
        18,
        144,
        147,
        120,
        90,
        117,
        180,
        86,
        131,
        250,
        252,
        65,
        132,
        246,
        100,
        121,
        52,
        53,
        28,
        172,
        122,
        133,
        14,
        115,
        120,
        114,
        1,
        231,
        36,
        137,
        37,
        158,
        218,
        127,
        101,
        188,
        175,
        135,
        147,
        25,
        140,
        219,
        117,
        21,
        182,
        224,
        48,
        199,
        8,
        248,
        89,
      ],
      "publicKey": "(2048 bit)",
    },
  },
  "subjectUniqueID": undefined,
  "validity": Object {
    "notAfter": 2017-08-21T05:27:41.000Z,
    "notBefore": 2012-08-22T05:27:41.000Z,
  },
  "version": 0,
}
`;
